bench/db_bench.ml

```diff
@@ -0,0 +1,267 @@
+    let sorted_uniq =
+      List.sort_uniq (fun a b -> compare (fst a) (fst b)) random
+    in
+    let nb_entries = List.length sorted_uniq in
+    print_results (write rw sorted_uniq) nb_entries;
+    Index.flush rw;
+    write_amplif (file_size (root // "fill_random")) nb_entries
+
+  let read_random r = print_results (read r random) nb_entries
+
+  let read_seq r = print_results (read r sorted) nb_entries
+
+  let read_seq_rev r = print_results (read r (List.rev sorted)) nb_entries
+
+  let close rw = Index.close rw
+end
+
+module Lmdb = struct
+  open Lmdb
+
+  let root = "/tmp"
+
+  let print_results = print_results "lmdb"
+
+  let cleanup () =
+    let files = [ root // "data.mdb"; root // "lock.mdb" ] in
+    ListLabels.iter files ~f:(fun fn -> Sys.(if file_exists fn then remove fn))
+
+  let file_size () =
+    let offset = get_offset (root // "data.mdb") in
+    if offset = 0 then failwith ("Error opening files in " ^ root ^ "/data.mdb");
+    offset
+
+  let fail_on_error f =
+    match f () with Ok _ -> () | Error err -> failwith (string_of_error err)
+
+  let flags = [ Lmdb.NoRdAhead; Lmdb.NoSync ]
+
+  let mapsize = 409_600_000_000L
+
+  let get_wtxn dir =
+    cleanup ();
+    opendir dir ~mapsize ~flags 0o644 >>= fun env ->
+    create_rw_txn env >>= fun txn ->
+    opendb txn >>= fun ddb -> Ok ((txn, ddb), env)
+
+  let write (txn, ddb) ls () =
+    List.iter
+      (fun (k, v) -> fail_on_error (fun () -> Lmdb.put_string txn ddb k v))
+      ls
+
+  let read (txn, ddb) ls () =
+    List.iter
+      (fun (k, _) ->
+        ignore (Bigstring.to_string (R.get_ok (Lmdb.get txn ddb k))))
+      ls
+
+  let write_random () =
+    get_wtxn root >>| fun (rw, env) ->
+    print_results (write rw random) nb_entries;
+    fail_on_error (fun () -> sync env);
+    write_amplif (file_size ()) nb_entries;
+    (rw, env)
+
+  let write_seq () =
+    get_wtxn root >>| fun (rw, env) ->
+    print_results (write rw sorted) nb_entries;
+    fail_on_error (fun () -> sync env);
+    write_amplif (file_size ()) nb_entries;
+    closedir env
+
+  let write_sync () =
+    get_wtxn root >>| fun (rw, env) ->
+    let write (txn, ddb) env ls () =
+      List.iter
+        (fun (k, v) ->
+          fail_on_error (fun () ->
+              Lmdb.put_string txn ddb k v >>= fun () -> sync env))
+        ls
+    in
+    print_results (write rw env random) nb_entries;
+    write_amplif (file_size ()) nb_entries;
+    closedir env
+
+  let overwrite rw =
+    let sorted_uniq =
+      List.sort_uniq (fun a b -> compare (fst a) (fst b)) random
+    in
+    let nb_entries = List.length sorted_uniq in
+    print_results (write rw sorted_uniq) nb_entries;
+    write_amplif (file_size ()) nb_entries
+
+  let read_random r = print_results (read r random) nb_entries
+
+  let read_seq r = print_results (read r sorted) nb_entries
+
+  let read_seq_rev r = print_results (read r (List.rev sorted)) nb_entries
+
+  let close env = closedir env
+end
+
+let () =
```

Comment:
I think it's worth putting this in a `main` function and then having `let () = main ()` at the top-level, in case we want to invoke this without calling the executable.